tools/odrive/utils.py

```diff
@@ -118,11 +118,106 @@ def did_close(evt):
     plot_t = threading.Thread(target=plot_data)
     plot_t.daemon = True
     plot_t.start()
-    
 
     return cancellation_token;
     #plot_data()
 
+
+class BulkCapture:
+    '''
+    Asynchronously captures a bulk set of data when instance is created.
+
+    get_var_callback: a function that returns the data you want to collect (see the example below)
+    data_rate: Rate in hz
+    length: Length of time to capture in seconds
+
+    Example Usage:
+        capture = BulkCapture(lambda :[odrv0.axis0.encoder.pos_estimate, odrv0.axis0.controller.pos_setpoint])
+        # Do stuff while capturing (like sending position commands)
+        capture.event.wait() # When you're done doing stuff, wait for the capture to be completed.
+        print(capture.data) # Do stuff with the data
+        capture.plot_data() # Helper method to plot the data
+    '''
+
+    def __init__(self,
+                 get_var_callback,
+                 data_rate=500.0,
+                 length=2.0):
+        from threading import Event, Thread
+        import pandas as pd
+
+        self.event = Event()
+        def loop():
+            vals = []
+            start_time = time.monotonic()
+            total_samples = int(length * data_rate)
+            period = 1.0 / data_rate
+            for i in range(total_samples):
+                try:
+                    data = get_var_callback()
+                except Exception as ex:
+                    print(str(ex))
+                    print("Waiting 1 second before next data point")
+                    time.sleep(1)
+                    continue
+                relative_time = time.monotonic() - start_time
+                vals.append([relative_time] + data)
+                time.sleep(period - (relative_time % period)) # this ensures consistently timed samples
+            self.data = pd.DataFrame(vals) # A lock is not really necessary due to the event
```

Comment:
In order to keep dependencies low I think we should expose either the plain python array here or use numpy (which we already use in other places) instead of pandas.

The plotting command below would then become `plt.plot(data[:,0], data[:,1:])`.